### ISSUE TITLE
Output often cutting off midway through

### ISSUE THREAD

**Issue by parallax4** [calzoneman/aeiou#18]
@calzoneman I really appreciate the free TTS service you're running and am not expecting anything here, but I just wanted to report an issue I've frequently been running into. Often output from text inputs that are longer than a few seconds will just suddenly, randomly cut off ~5 - 10 seconds in. Sometimes trying again by adding or removing a letter totally fixes it (I assume by generating from scratch again instead of using the cut-off cached copy) and sometimes it gets to the end, or sometimes it'll cut off at a different point.

Is there anything I can do to reduce the chance of this occurring? Thank you.

**Comment by calzoneman** [comment]
Do you have an example of an input that causes this problem?

There are certain inputs that can cause DECtalk to crash partway through due to programming errors in DECtalk itself.  I've attempted to patch a few of them but there could be more that you're running into.

**Comment by parallax4** [comment]
One input will cause it and minutes/hours later no longer cause it (perhaps due to cache expiry). For example, about an hour ago this WAV cut off after 6 seconds and was 6 seconds long (I have a copy of it I can upload if needed, but it's nothing interesting - just abruptly cuts off without anything else abnormal), while now it generates correctly and is 16 seconds long, as it's supposed to be. The filename/URL is exactly the same for both the truncated and correct versions.

Input:

```
[gley<1000,12> gley<1000,24> gley<1000,19> _<1000> gley<1000,17> gley<1000,24> gley<1000,12> _<1000> gley<1000,12> gley<1000,17> gley<1000,24> gley<1000,26> gley<1000,24> gley<1000,19> gley<1000,17>]
```

File: https://tts.cyzon.us/files/00908af127b734e03c8c5115bb90e65e.wav

I know this isn't a very helpful example since this currently generates correctly - it only generated incorrectly about an hour ago, at around 4:08 UTC. It seems to be a sporadic and inconsistent problem. Anything I can give you to test with may suddenly start working when you try it, so it's tough.

Just changing one letter of text in the input can make things suddenly work (or just cut off a bit before or after when the original one did), and the same exact input will also often start to work again minutes/hours later.

It happens on all kinds of inputs. Both the phoneme stuff and regular plaintext, with or without commands. The only pattern I've noticed is just that it seems to apply to any outputs that (should) last more than a few seconds. I don't think I've ever heard a 3 second one get cut off. But many 8+ second ones can become significantly shorter.

**Comment by calzoneman** [comment]
OK, I pulled the logs for that filename and I'm pretty sure I know what's going on.

There were a bunch of requests to render that same string within the same second.  The webserver handles these requests in three tiers:

1. If the rendered file already exists, then redirect immediately.
2. If there is another request already rendering this string, then wait for it to finish and then redirect to the result.
3. Queue a new task to render the string.

There is an edge case here, if the task has *started* (but not finished) rendering the file, then a request will observe that the file exists and redirect to it, even though it hasn't finished yet.  This could theoretically happen to any request, but is more likely to happen if DECtalk is rendering longer files (larger time window to encounter the bug).

The simple fix here is to reverse the order of (1) and (2) -- if a task is in progress we should wait for it to finish regardless of whether the file exists, since it may not be a complete file yet.  It may also be a good idea to just write all files to a temporary location and then copy them to the public-facing folder when they're finished, to avoid corrupted files due to DECtalk crashes etc.

As an aside, if you're using this API to distribute text to speech to multiple clients, it would be preferable if you could make one call to the API and then distribute the audio yourself; while I have caching in place, the API sees bursty traffic due to dozens of clients making the same requests at the same time.

**Comment by parallax4** [comment]
Ah, I understand. I'll change it so our server downloads and hosts the file instead of having all of the clients load it from your server. Thanks.

**Comment by calzoneman** [comment]
Re-opening to track fixing the above.

**Comment by calzoneman** [comment]
Pushed & purged cache, let me know if this problem reappears.

**Comment by parallax4** [comment]
@calzoneman Sorry for all the traffic. Starting from when you requested it, I've been working on reimplementing it on the server so there'll only be one request per invocation, but there's a lot of client-side code I still need to port and I haven't gotten a chance to complete it all yet. We'll disable the feature until I've gotten a chance to finish moving it.

**Comment by calzoneman** [comment]
I'm not sure what prompted this sudden comment, but if it's related to receiving 429s since yesterday, that's a slightly different issue.  Yesterday, a small number of IPs was generating 4+ requests/sec with unique inputs on each request (normal traffic is <0.5 req/sec on average), which overloaded the small resources I've provisioned for DECtalk and it wasn't clear whether those inputs were legitimate or a runaway script.  I applied a flat rate limit for any IP that exceeds 30 requests/min on average, but I plan to add some better rate limiting later.

I plan to publish some API usage guidelines, including reaching out to discuss any usages that plan to exceed 15 requests/min.  As an aside, until I get around to implementing some kind of API key authentication, it would be useful if you customized your HTTP client's User-Agent to something unique and possibly including your contact information, that way if I see traffic from specific applications I can bucket those and follow up more easily.
